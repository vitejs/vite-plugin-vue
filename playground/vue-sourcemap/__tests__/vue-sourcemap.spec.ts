import { URL } from 'node:url'
import { describe, expect, test } from 'vitest'
import {
  extractSourcemap,
  formatSourcemapForSnapshot,
  isBuild,
  isServe,
  page,
  serverLogs,
} from '~utils'

describe.runIf(isServe)('serve:vue-sourcemap', () => {
  const getStyleTagContentIncluding = async (content: string) => {
    const styles = await page.$$('style')
    for (const style of styles) {
      const text = await style.textContent()
      if (text.includes(content)) {
        return text
      }
    }
    throw new Error('Style not found: ' + content)
  }

  test('js', async () => {
    const res = await page.request.get(new URL('./Js.vue', page.url()).href)
    const js = await res.text()
    const map = extractSourcemap(js)
    expect(formatSourcemapForSnapshot(map)).toMatchSnapshot('serve-js')
  })

  test('ts', async () => {
    const res = await page.request.get(new URL('./Ts.vue', page.url()).href)
    const js = await res.text()
    const map = extractSourcemap(js)
    expect(formatSourcemapForSnapshot(map)).toMatchSnapshot('serve-ts')
  })

  test('css', async () => {
    const css = await getStyleTagContentIncluding('.css ')
    const map = extractSourcemap(css)
    expect(formatSourcemapForSnapshot(map)).toMatchSnapshot('serve-css')
  })

  test('css module', async () => {
    const css = await getStyleTagContentIncluding('._css-module_')
    const map = extractSourcemap(css)
    expect(formatSourcemapForSnapshot(map)).toMatchSnapshot('serve-css-module')
  })

  test('css scoped', async () => {
    const css = await getStyleTagContentIncluding('.css-scoped[data-v-')
    const map = extractSourcemap(css)
    expect(formatSourcemapForSnapshot(map)).toMatchSnapshot('serve-css-scoped')
  })

  test('sass', async () => {
    const css = await getStyleTagContentIncluding('.sass ')
    const map = extractSourcemap(css)
    expect(formatSourcemapForSnapshot(map)).toMatchSnapshot('serve-sass')
  })

  test('sass with import', async () => {
    const css = await getStyleTagContentIncluding('.sass-with-import ')
    const map = extractSourcemap(css)
    expect(formatSourcemapForSnapshot(map)).toMatchSnapshot(
      'serve-sass-with-import',
    )
  })

  test('less with additionalData', async () => {
    const css = await getStyleTagContentIncluding('.less ')
    const map = extractSourcemap(css)
    expect(formatSourcemapForSnapshot(map)).toMatchSnapshot(
      'serve-less-with-additionalData',
    )
  })

  test('src imported', async () => {
    const css = await getStyleTagContentIncluding('.src-import[data-v-')
    const map = extractSourcemap(css)
    expect(formatSourcemapForSnapshot(map)).toMatchSnapshot(
      'serve-src-imported',
    )
  })

  test('src imported sass', async () => {
    const css = await getStyleTagContentIncluding('.src-import-sass[data-v-')
    const map = extractSourcemap(css)
    expect(formatSourcemapForSnapshot(map)).toMatchSnapshot(
      'serve-src-imported-sass',
    )
  })

  test('no script', async () => {
    const res = await page.request.get(
      new URL('./NoScript.vue', page.url()).href,
    )
    const js = await res.text()
    const map = extractSourcemap(js)
    expect(formatSourcemapForSnapshot(map)).toMatchSnapshot('serve-no-script')
  })

  test('no template', async () => {
    const res = await page.request.get(
      new URL('./NoTemplate.vue', page.url()).href,
    )
    const js = await res.text()
    const map = extractSourcemap(js)
    expect(formatSourcemapForSnapshot(map)).toMatchSnapshot('serve-no-template')
  })
})

test.runIf(isBuild)('should not output sourcemap warning (#4939)', () => {
  serverLogs.forEach((log) => {
    expect(log).not.toMatch('Sourcemap is likely to be incorrect')
  })
})
