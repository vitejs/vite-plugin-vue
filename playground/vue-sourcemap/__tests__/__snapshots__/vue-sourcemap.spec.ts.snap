// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`serve:vue-sourcemap > css > serve-css 1`] = `
{
  "mappings": ";AAQA;EACE,UAAU;AACZ",
  "sources": [
    "/root/Css.vue",
  ],
  "sourcesContent": [
    "<template>
  <p class=\\"css\\">&lt;css&gt;</p>
  <p :class=\\"$style['css-module']\\">&lt;css&gt; module</p>
  <p class=\\"css-scoped\\">&lt;css&gt; scoped</p>
  <p class=\\"css-scoped-nested\\">&lt;css&gt; scoped with nested</p>
</template>

<style>
.css {
  color: red;
}
</style>

<style module>
.css-module {
  color: red;
}
</style>

<style scoped>
.css-scoped {
  color: red;
}
</style>

<style scoped>
.css-scoped-nested {
  color: red;
  .dummy {
    color: green;
  }
  font-weight: bold;
}
</style>
",
  ],
  "version": 3,
}
`;

exports[`serve:vue-sourcemap > css module > serve-css-module 1`] = `
{
  "mappings": ";AAcA;EACE,UAAU;AACZ",
  "sources": [
    "/root/Css.vue",
  ],
  "sourcesContent": [
    "<template>
  <p class=\\"css\\">&lt;css&gt;</p>
  <p :class=\\"$style['css-module']\\">&lt;css&gt; module</p>
  <p class=\\"css-scoped\\">&lt;css&gt; scoped</p>
  <p class=\\"css-scoped-nested\\">&lt;css&gt; scoped with nested</p>
</template>

<style>
.css {
  color: red;
}
</style>

<style module>
.css-module {
  color: red;
}
</style>

<style scoped>
.css-scoped {
  color: red;
}
</style>

<style scoped>
.css-scoped-nested {
  color: red;
  .dummy {
    color: green;
  }
  font-weight: bold;
}
</style>
",
  ],
  "version": 3,
}
`;

exports[`serve:vue-sourcemap > css scoped > serve-css-scoped 1`] = `
{
  "mappings": ";AAoBA;EACE,UAAU;AACZ",
  "sources": [
    "/root/Css.vue",
  ],
  "sourcesContent": [
    "<template>
  <p class=\\"css\\">&lt;css&gt;</p>
  <p :class=\\"$style['css-module']\\">&lt;css&gt; module</p>
  <p class=\\"css-scoped\\">&lt;css&gt; scoped</p>
  <p class=\\"css-scoped-nested\\">&lt;css&gt; scoped with nested</p>
</template>

<style>
.css {
  color: red;
}
</style>

<style module>
.css-module {
  color: red;
}
</style>

<style scoped>
.css-scoped {
  color: red;
}
</style>

<style scoped>
.css-scoped-nested {
  color: red;
  .dummy {
    color: green;
  }
  font-weight: bold;
}
</style>
",
  ],
  "version": 3,
}
`;

exports[`serve:vue-sourcemap > js > serve-js 1`] = `
{
  "mappings": "AAKA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;;;;;;AAGP;AACd,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;;;;;;;;;;;wBARlB,oBAAiB,WAAd,MAAU",
  "sources": [
    "Js.vue",
  ],
  "sourcesContent": [
    "<template>
  <p>&lt;js&gt;</p>
</template>

<script>
console.log('script')
</script>

<script setup>
console.log('setup')
</script>
",
  ],
  "version": 3,
}
`;

exports[`serve:vue-sourcemap > less with additionalData > serve-less-with-additionalData 1`] = `
{
  "mappings": "AAKA;EACE",
  "sources": [
    "/root/Less.vue",
  ],
  "sourcesContent": [
    "<template>
  <p class=\\"less\\">&lt;less&gt; with additionalData</p>
</template>

<style lang=\\"less\\">
.less {
  color: @color;
}
</style>
",
  ],
  "version": 3,
}
`;

exports[`serve:vue-sourcemap > no script > serve-no-script 1`] = `
{
  "mappings": ";;;wBACE,oBAAwB,WAArB,aAAiB",
  "sourceRoot": "",
  "sources": [
    "NoScript.vue",
  ],
  "sourcesContent": [
    "<template>
  <p>&lt;no-script&gt;</p>
</template>
",
  ],
  "version": 3,
}
`;

exports[`serve:vue-sourcemap > no template > serve-no-template 1`] = `
{
  "mappings": "AACA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;;;;;;AAGP;AACd,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;;;;;;;",
  "sources": [
    "NoTemplate.vue",
  ],
  "sourcesContent": [
    "<script>
console.log('script')
</script>

<script setup>
console.log('setup')
</script>
",
  ],
  "version": 3,
}
`;

exports[`serve:vue-sourcemap > sass > serve-sass 1`] = `
{
  "mappings": "AAKA;EACE",
  "sources": [
    "/root/Sass.vue",
  ],
  "sourcesContent": [
    "<template>
  <p class=\\"sass\\">&lt;sass&gt;</p>
</template>

<style lang=\\"sass\\">
.sass
  color: red
</style>
",
  ],
  "version": 3,
}
`;

exports[`serve:vue-sourcemap > sass with import > serve-sass-with-import 1`] = `
{
  "mappings": "AAAA;EACE;;ACOF;EACE",
  "sources": [
    "/root/sassWithImportImported.sass",
    "/root/SassWithImport.vue",
  ],
  "sourcesContent": [
    ".sass-with-import-imported
  color: red
",
    "<template>
  <p class=\\"sass-with-import\\">&lt;sass&gt; with import</p>
  <p class=\\"sass-with-import-imported\\">&lt;sass&gt; with import (imported)</p>
</template>

<style lang=\\"sass\\">
@import './sassWithImportImported'

.sass-with-import
  color: red
</style>
",
  ],
  "version": 3,
}
`;

exports[`serve:vue-sourcemap > src imported > serve-src-imported 1`] = `
{
  "mappings": "AAAA;EACE,UAAU;AACZ",
  "sources": [
    "/root/src-import/src-import.css",
  ],
  "sourcesContent": [
    ".src-import {
  color: red;
}
",
  ],
  "version": 3,
}
`;

exports[`serve:vue-sourcemap > src imported sass > serve-src-imported-sass 1`] = `
{
  "mappings": "AAAA;EACE;;ACCF;EACE",
  "sources": [
    "/root/src-import/src-import-imported.sass",
    "/root/src-import/src-import.sass",
  ],
  "sourcesContent": [
    ".src-import-sass-imported
  color: red
",
    "@import './src-import-imported'

.src-import-sass
  color: red
",
  ],
  "version": 3,
}
`;

exports[`serve:vue-sourcemap > ts > serve-ts 1`] = `
{
  "mappings": ";AAKA,QAAQ,IAAI,WAAW;;;;;AAIvB,YAAQ,IAAI,UAAU;;;;;;;;uBARpB,oBAAiB,WAAd,MAAU",
  "sources": [
    "Ts.vue",
  ],
  "sourcesContent": [
    "<template>
  <p>&lt;ts&gt;</p>
</template>

<script lang=\\"ts\\">
console.log('ts script')
</script>

<script lang=\\"ts\\" setup>
console.log('ts setup')
</script>
",
  ],
  "version": 3,
}
`;
